compiler/codegen/codegen.cpp

```diff
@@ -2132,21 +2177,58 @@ void codegen(void) {
       lastSlash++;
     }
 
-    // copy from that slash onwards into the executableFilename,
-    // saving space for a `\0` terminator
-    if (strlen(lastSlash) >= sizeof(executableFilename)) {
-      INT_FATAL("input filename exceeds executable filename buffer size");
+    // "Executable" name should be given a "lib" prefix in library compilation,
+    // and just the main module name in normal compilation.
+    if (fLibraryCompile) {
+      // If the header name isn't set either, don't use the prefix version
+      if (libmodeHeadername[0] == '\0') {
+        // copy from that slash onwards into the libmodeHeadername,
+        // saving space for a `\0` terminator
+        if (strlen(lastSlash) >= sizeof(libmodeHeadername)) {
+          INT_FATAL("input filename exceeds header filename buffer size");
+        }
+        strncpy(libmodeHeadername, lastSlash, sizeof(libmodeHeadername)-1);
+        libmodeHeadername[sizeof(libmodeHeadername)-1] = '\0';
+        // remove the filename extension from the library header name.
+        char* lastDot = strrchr(libmodeHeadername, '.');
+        if (lastDot == NULL) {
+          INT_FATAL(mainMod,
+                    "main module filename is missing its extension: %s\n",
+                    libmodeHeadername);
+        }
+        *lastDot = '\0';
+      }
+      if (strlen(lastSlash) >= sizeof(executableFilename) - 3) {
+        INT_FATAL("input filename exceeds executable filename buffer size");
+      }
+      strncpy(executableFilename, "lib", 3);
+      strncat(executableFilename, lastSlash, sizeof(executableFilename)-4);
+      executableFilename[sizeof(executableFilename)-1] = '\0';
+
+    } else {
+      // copy from that slash onwards into the executableFilename,
+      // saving space for a `\0` terminator
+      if (strlen(lastSlash) >= sizeof(executableFilename)) {
+        INT_FATAL("input filename exceeds executable filename buffer size");
+      }
+      strncpy(executableFilename, lastSlash, sizeof(executableFilename)-1);
+      executableFilename[sizeof(executableFilename)-1] = '\0';
     }
-    strncpy(executableFilename, lastSlash, sizeof(executableFilename)-1);
-    executableFilename[sizeof(executableFilename)-1] = '\0';
 
-    // remove the filename extension
+    // remove the filename extension from the executable filename
     char* lastDot = strrchr(executableFilename, '.');
     if (lastDot == NULL) {
       INT_FATAL(mainMod, "main module filename is missing its extension: %s\n",
                 executableFilename);
     }
     *lastDot = '\0';
+
+  }
+
+  // If we're in library mode and the executable name was set but the header
+  // name wasn't, use the executable name for the header name as well
+  if (fLibraryCompile && libmodeHeadername[0] == '\0') {
+    strncpy(libmodeHeadername, executableFilename, sizeof(executableFilename));
```

Comment:
Can all this computing-library-file-names stuff go into a new function? Just because codegen() is really big doesn't mean we should make it worse ;)